### ISSUE TITLE
Input Events wait for Update Cycle - inconsistent timing

### ISSUE THREAD

**Issue by najak3d** [urho3d/Urho3D#2850]
It's vital for input events to be fired in real-time, so that they are not impacted/tainted by a Render/Update hiccup, or a slow render.   The Input thread is the topmost important thread, with no hiccups, to guarantee that the system response to User Input is consistent, regardless of Rendering Frame rate (or hiccups).

Urho's Input events seem to be synchronized with the Update cycle, which means the lag/intervals can be inconsistent and higher risk of lag.    Can we please provide an option to configure Input events to happen real-time without lag, rather than being sync'ed to Update cycle?

For sake of understanding, here is an extreme example:
1. If the mouse is moved 1 inch in 10 msec -- that's a very fast twitch.   In order to detect just how fast is this twitch, it requires an input sampling rate of at least 100 Hz.   
2. If the input thread hiccups (along with the rendering thread) and let's say it takes 1 Second between updates -- now it appears instead that the mouse moved just 1 inch in 1 Second, which appears to be a 100x slower than it actually was.

NOTE: That on most desktops, like Windows, the default Mouse Behavior is to "move much further distances for fast mouse movement" -- so that the movement of the mouse is not linearly-proportionate to the mouse distance moved -- but rather is very much a function of the "speed of the motion", not just the distance.   Fast reliable sampling rate is required to provide this type of accelerated response.

It is of utmost importance that the application responds to User Input CONSISTENTLY regardless of Rendering Frame rate, and therefore it's of utmost importance to maintain the fastest-possible sampling rate of User Inputs, with the minimal amount of lag.

In short -- please provide an option for the developer to enable "Immediate Input Events" on a separate UI thread, so that we can provide this benefit to our users.   Thanks!

===
NOTE: It's probably still a good idea to always provide the user with an Update-synchronized Input event too (for those who enjoy the convenience of knowing the event is firing on the Update thread).   There is no reason that both queues of input events cannot be active at the same time.

**Comment by eugeneko** [comment]
> an option for the developer to enable "Immediate Input Events" on a separate UI thread, so that we can provide this benefit to our users

The question is: how to we _get_ input events on separate thread? I am not sure SDL allows it.

**Comment by najak3d** [comment]
> > an option for the developer to enable "Immediate Input Events" on a separate UI thread, so that we can provide this benefit to our users
> 
> The question is: how to we _get_ input events on separate thread? I am not sure SDL allows it.

The would be very surprising.  When we were hosting Avalonia inside of an Urho Sprite (works great), the input processor for Avalonia was receiving mouse updates at about 500 Hz.  This was on Windows.

I can't imagine that SDL doesn't provide native access to immediate Input events as they arise.   I figured Urho must just be doing a Polling method, or a "Last event" method -- so that when the Urho Update thread triggered another update, it would simply either: 
(a) Poll for current mouse/keyboard state, OR 
(b) Simply grab the "Last native event" captured (and forget all previous events).

XNA worked like (a) above -- it just polled, OnUpdate cycle for "current Keyboard/Mouse" states.   Very inadequate, so when we were using XNA, we gathered our inputs straight from the WinForm. 

It would be highly deficient of SDL if they didn't provide inputs at some guaranteed/fast rate.

We ran a test with current Urho, and the Input events were firing In-Synch with the Update calls (one for one) -- so even if we had a long-Update, the Input would then slow down to match the Frame rate, 1:1 basis.   This is just wrong, and unreliable.

**Comment by eugeneko** [comment]
> I can't imagine that SDL doesn't provide native access to immediate Input events as they arise

As far as I understand, SDL requires event processing in the same thread as window creation.
So, "input thread" should also be accessible in `Graphics` and maybe somewhere else, I'm not sure how well SDL can deal with calls from multiple threads.

**Comment by najak3d** [comment]
> As far as I understand, SDL requires event processing in the same thread as window creation. So, "input thread" should also be accessible in `Graphics` and maybe somewhere else, I'm not sure how well SDL can deal with calls from multiple threads.

It's fine to "read" the event(s) (Plural) from the main rendering thread, so long as those events can be plural.  In short, if your rendering loop hiccups to 100 msec, there may be 10-20 mouse move events that occurred during this hiccup.   If we can have access to "all 20 mouse events" inside the main thread -- this works fine too.   What we do NOT want is to ONLY have access to the "accumulated/aggregated" singular mouse-move event (that just shows the "net movement in the last 100 msec").

We need access to the fine-granularity of input events (with time stamps on each) in order to produce a reliable/consistent User Response (not impacted by rendering speeds, or hiccups).   Inside the rendering loop, we'd just enumerate the individual mouse move events one-by-one to the end.

If SDL hides these rapid individual events, then SDL is defective (and another means for obtaining user inputs is needed).   But if SDL permits access to these events, then likewise, so should Urho3D.

I've been googling this -- and at this point, to me, it doesn't look good for SDL itself.

**Comment by ToolmakerSteve** [comment]
@eugeneko - what surprises me is that events aren't getting *queued*. In other words, even if our app code can't see those events until we return from OnUpdate, I would expect interrupt-priority code to exist, that could queue events. If that existed, then after our OnUpdate returns, our mouse event handler would get called several times. Before the next OnUpdate.

That would make it possible for us to internally cache those events. (that occurred between two OnUpdates). Assuming SDL events contain a timestamp, we could then determine exactly what the user intended, in their input actions.

In the case where I encountered the problem, its an "editor", and the person is drawing a freehand curve, that gets turned into a wall. There are moments where adding a wall segment is slow - trying to figure out how to not "lose" the information about the path the mouse was on - currently the result is unacceptable.

Could do this easily many years ago on Windows XP using GDI. Of course that was just 2D drawing. Now I wish to find a good way to do the equivalent in 3D.

And najak has other situations where accurate, continuous, user input is vital.

**Comment by ToolmakerSteve** [comment]
**UPDATE** - TODAY, I AM getting several (two or three) mouse event handler calls on some frames.

This is encouraging. I'll see if (the minimum time delay between mouse events) improves further, after I've moved some "prep" code out of OnUpdate, to a background thread.

I have no idea why I did not see those multiple events in my testing yesterday.

-----------------------

This raises a new question: Is there a setting somewhere that could be increased, to try to get more frequent mouse events when desired?
